### PR TITLE
Improve handling of transaction index

### PR DESCRIPTION
**Pull request by zathras-crypto**
Various feedback has been received around support for the txindex flag and our recent change to default it back to match Bitcoin Core.

I've defended this decision as I believe it is the right thing to do, but it has caused some users problems whom were relying on our hardcoded txindex value.

Two issues present on this (that I've seen):
1) Users who struggle to update a configuration file manually
2) Users who do not know where their configuration file is

Number 2) actually surprised me, but I've assisted a couple of users who for example have run Bitcoin Core before and chosen a custom datadir, subsequently uninstalled it and then when using OmniCore cannot find where the config file is located.

This small PR serves to address those two points by providing the configuration file location directly within the txindex disabled error message, and by offering to update it automatically on behalf of the user.

Feedback welcomed :)

Thanks
Z

image

### CODE REVIEW

**Comment by dexX7** [comment]
I had huge reservation against those types of changes, but you convinced me immediately, not only with the picture, but also with the very un-intrusive code change - I really like it!

I'm going to test it shortly. Quick question in general: `"OmniCore"` or `"Omni Core"`? I'd prefer the later.

**Comment by zathras-crypto** [comment]
> Quick question

Ah yes, oops - yeah it should be Omni Core :)

File: src/init.cpp [hunk 1]
```diff
@@ -1118,13 +1118,35 @@ bool AppInit2(boost::thread_group& threadGroup)
...
-                "command line argument or add \"txindex=1\" to your client "
-                "configuration file."
-            ));
+        // ask the user if they would like us to modify their config file for them
+        std::string msg = _("Disabled transaction index detected.\n\n"
+                            "Omni Core requires an enabled transaction index. To enable "
+                            "transaction indexing, please use the \"-txindex\" option as "
+                            "command line argument or add \"txindex=1\" to your client "
+                            "configuration file within your data directory.\n\n"
+                            "Configuration file: "); // allow translation of main text body while still allowing differing config file string
```
Comment (dexX7) [inline src/init.cpp]: I've never worked with Transifex, but I think words and texts to translate have to be isolated, such as `"File"` vs. `"File: "`. A bit cumbersome though, hehe. Another example:

```c++
std::string msg = _("Disabled transaction index detected.") + "\n\n";
msg += _("Omni Core requires an enabled transaction index. To enable "
         "transaction indexing, please use the \"-txindex\" option as "
         "command line argument or add \"txindex=1\" to your client "
         "configuration file within your data directory.") + "\n\n";
msg += _("Configuration file") + ": ";
```

Comment (zathras-crypto) [inline src/init.cpp]: Ahh OK - I can update to work with that - I had thought it was entire strings that were translated

Comment (zathras-crypto) [inline src/init.cpp]: If you look at any of the localization files, this is what I see (& hence thought whole strings were translated):
```
<message>
        <source>Double-click to edit address or label</source>
        <translation>Для того, чтобы изменить адрес или метку, дважды кликните по изменяемому объекту</translation>
    </message>
```

Comment (zathras-crypto) [inline src/init.cpp]: Ahh, the penny drops - transifex creates these localization files perhaps?

Comment (dexX7) [inline src/init.cpp]: Yes, all those messages to translate are extracted and then processed. :)

**Comment by dexX7** [comment]
Still building locally, but just to outline what I'm looking for: odd scenarios. E.g.: what happens, if the user doesn't use the QT client at all?

**Comment by dexX7** [comment]
Another one, but I think there is no nice way to handle it, and it's too improbable anyway: it's entirely possible that an user has explicitly `txindex=0` in config.

**Comment by zathras-crypto** [comment]
> E.g.: what happens, if the user doesn't use the QT client at all?

Yeah I tested that to make sure, it's not perfect but it doesn't seem problematic.  Below is output, no input is requested, it just shuts down.

```
zathras@coredev01:~/github/build/omnicore$ src/bitcoind --server --daemon --testnet
Bitcoin server starting
zathras@coredev01:~/github/build/omnicore$ : Disabled transaction index detected.

Omni Core requires an enabled transaction index. To enable transaction indexing, please use the "-txindex" option as command line argument or add "txindex=1" to your client configuration file within your data directory.

Configuration file: /home/zathras/.bitcoin/bitcoin.conf

Would you like OmniCore to attempt to update your configuration file accordingly?
Error: Please add txindex=1 to your configuration file manually.

OmniCore will now shutdown
2015-05-06 02:05:13 Omni Core shutdown completed
```

File: src/init.cpp [hunk 2]
```diff
@@ -1118,13 +1118,35 @@ bool AppInit2(boost::thread_group& threadGroup)
...
+        bool fRet = uiInterface.ThreadSafeMessageBox(msg, "", CClientUIInterface::MSG_ERROR | CClientUIInterface::BTN_ABORT);
+        if (fRet) {
+            // add txindex=1 to config file in GetConfigFile()
+            boost::filesystem::path configPathInfo = GetConfigFile();
+            FILE *fp = fopen(configPathInfo.string().c_str(), "a");
+            if (!fp) {
+                std::string failMsg = _("Unable to update configuration file at:\n");
+                failMsg += GetConfigFile().string() + "\n\n";
+                failMsg += _("The file may be write protected or you may not have the required permissions to edit it.\n");
+                failMsg += _("Please add txindex=1 to your configuration file manually.\n\nOmni Core will now shutdown");
```
Comment (dexX7) [inline src/init.cpp]: Missing "." at the end.

File: src/init.cpp [hunk 3]
```diff
@@ -1118,13 +1118,35 @@ bool AppInit2(boost::thread_group& threadGroup)
...
+                return InitError(failMsg);
+            }
+            fprintf(fp, "\ntxindex=1\n");
+            fflush(fp);
+            fclose(fp);
+            return InitError(_("Your configuration file has been updated.\n\n"
+                               "Omni Core will now shutdown - please restart the client for your new configuration to take effect"));
```
Comment (dexX7) [inline src/init.cpp]: Missing "." at the end.

File: src/init.cpp [hunk 4]
```diff
@@ -1118,13 +1118,35 @@ bool AppInit2(boost::thread_group& threadGroup)
...
+        } else {
+            return InitError(_("Please add txindex=1 to your configuration file manually.\n\nOmni Core will now shutdown"));
```
Comment (dexX7) [inline src/init.cpp]: Missing "." at the end.

**Comment by zathras-crypto** [comment]
> Another one, but I think there is no nice way to handle it, and it's too improbable anyway: it's entirely possible that an user has explicitly txindex=0 in config.

Yeah I tested that too, sadly it looks like a second directive does not override the first.  So:
```
txindex=0
txindex=1
```
inside the conf results in a disabled transaction index.  I did toy around with pulling the file into a stream and then manipulating it but it got quite 'clunky' so I figured this would cover 99% of cases.

**Comment by dexX7** [comment]
I really love it, haha! :)

> I did toy around with pulling the file into a stream and then manipulating it but it got quite 'clunky' so I figured this would cover 99% of cases.

Let's keep it simple.

One last question: `\ntxindex=1\n` is added. Do we need the second `\n`?

**Comment by zathras-crypto** [comment]
> One last question: \ntxindex=1\n is added. Do we need the second \n?

I wasn't 100% sure whether it was safe to leave out a newline across all platforms so added it for good measure.

**Comment by dexX7** [comment]
> I wasn't 100% sure whether it was safe to leave out a newline across all platforms so added it for good measure.

Good point. This actually brings up another potential pitfall:

Unix uses `"\n"`, Windows `"\r\n"` and Mac `"\r"` for a new line.

It was suggested to open the file in `"at"` mode, whereby `"t"` stands for text, which appearingly encourages `fprintf` to "translate" the new line properly. I tested it locally (Ubuntu) and it still works.

I propose the following:
```diff
diff --git a/src/init.cpp b/src/init.cpp
index 1725e6e..638e3df 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1131,7 +1131,7 @@ bool AppInit2(boost::thread_group& threadGroup)
         if (fRet) {
             // add txindex=1 to config file in GetConfigFile()
             boost::filesystem::path configPathInfo = GetConfigFile();
-            FILE *fp = fopen(configPathInfo.string().c_str(), "a");
+            FILE *fp = fopen(configPathInfo.string().c_str(), "at");
             if (!fp) {
                 std::string failMsg = _("Unable to update configuration file at");
                 failMsg += ":\n" + GetConfigFile().string() + "\n\n";
@@ -1139,7 +1139,7 @@ bool AppInit2(boost::thread_group& threadGroup)
                 failMsg += _("Please add txindex=1 to your configuration file manually.\n\nOmni Core will now shutdown.");
                 return InitError(failMsg);
             }
-            fprintf(fp, "\ntxindex=1\n");
+            fprintf(fp, "\ntxindex=1");
```

**Comment by zathras-crypto** [comment]
> Good point. This actually brings up another potential pitfall:

Ahhh the old carriage return line feed jazz hehe...  Cool will apply your patch :)

**Comment by zathras-crypto** [comment]
Just tested a clean install and allowed OmniCore to fix my conf (one didn't exist before) - don't think the CRLF is necessarily working correctly:

```
zathras@zathras-desktop:~/.bitcoin$ cat bitcoin.conf 

txindex=1zathras@zathras-desktop:~/.bitcoin$ 
```

**Comment by dexX7** [comment]
@zathras-crypto: isn't this expected?

No line break after the line, but one before?

**Comment by zathras-crypto** [comment]
> No line break after the line, but one before?

Yeah I see - I think I just misinterpreted what you were saying & implemented your suggestion directly

> I propose the following:
```
-            FILE *fp = fopen(configPathInfo.string().c_str(), "a");
+            FILE *fp = fopen(configPathInfo.string().c_str(), "at");
...
-            fprintf(fp, "\ntxindex=1\n");
+            fprintf(fp, "\ntxindex=1");
```

thinking the switch to text mode removed the need for a line break at the end - will add the last \n back in...

Thanks

**Comment by dexX7** [comment]
Sorry, I can't follow? I initially proposed only the line break before the line, to avoid ending up with a scattered config like this:

```
line

line

line
```

**Comment by zathras-crypto** [comment]
Without line break at the end:
```
zathras@zathras-desktop:~/.bitcoin$ cat bitcoin.conf 
existingconfigline1
existingconfigline2
existingconfigline3
existingconfigline4

txindex=1zathras@zathras-desktop:~/.bitcoin$ 
```

With line break at the end:

```
zathras@zathras-desktop:~/.bitcoin$ cat bitcoin.conf 
existingconfigline1
existingconfigline2
existingconfigline3
existingconfigline4

txindex=1
zathras@zathras-desktop:~/.bitcoin$ 
```

To me the second seems 'cleaner'

**Comment by dexX7** [comment]
True, this is cleaner.

**Pull request merged**